app/ui/src/main/java/de/mm20/launcher2/ui/launcher/search/SearchVM.kt

```diff
@@ -102,7 +105,7 @@ class SearchVM : ViewModel(), KoinComponent {
     }
 
     private var searchJob: Job? = null
-    fun search(query: String, forceRestart: Boolean = false) {
+    fun search(query: String, context: Context?, forceRestart: Boolean = false) {
```

Comment:
I don't like how you change my function signatures. Especially that context being nullable here screams ugly hack to me. Here is a better approach:

- Add a new class `UserLocationProvider`, inject context as a constructor parameter
- Add a Koin declaration for that class `single<UserLocationProvider>{ UserLocationProvider(androidContext()) }` where you provide the application context as a dependency
- Inject an instance of that class into `SearchVM` using Koin's `by inject()`

Reply:
So the context-instance doesn't change during runtime? I'll change that. 
Anyway, if there are any other changes that you don't like, tell me, no hard feelings. Many of these commits follow the "make it work, then make it right"-mentality anyway :)

Reply:
I usually try to avoid passing contexts around whenever possible. There's two reasons for this:

1. Contexts are an easy way to create memory leaks. You always need to make sure that you don't store activity context references in places that outlive that activity (or the activity can't be garbage collected when it is destroyed). That's the reason we don't store a direct reference to the context in our ViewModel. The exception is the application context, which lives as long as the application process is alive. This is what Koin injects when we call `androidContext`. Technically, there is `AndroidViewModel` which provides an application context in its constructor, but I consider this an antipattern two, because
2. Contexts are way too platform specific. Yes, we develop an Android app, but we should still try to avoid using them in public interfaces. Because
   1. It's more difficult to write unit tests if the method depends on a context (joke's on me who has written like two unit tests in total for this repo. I will add them eventually, I promise!)
   2. Once you start to require them, you will always be confronted with the question "from where do I get a context now?". And then you infect your entire code with it, because, well, the obvious solution is to make the user of that function require a context as well. We can see that in your code, you found yourself asking the exact same question and your solution was to make the context nullable.

So essentially my philosophy is that I don't use `Context`s in the public interface of a class[^1], except
- if we explicitely need an activity context (e.g. `SavableSearchable.launch`, need the context we launch from)
- in classes that are essentially data classes (e.g. `SavableSearchable.loadIcon`) – I don't want to store context references in data classes so I need to provide an external one.

In both cases, the context reference must be released after the function returns.

In all other cases, I keep the public interface context-free and keep it hidden as an implementation detail that is injected by Koin if needed.

[^1] I  say public interface because technically the cleanest solution to your problem would be to write an `interface UserLocationProvider` and a separate implementation `class UserLocationProviderImpl: UserLocationProvider`, but that's overkill if you only have one implementation. And if you ever find yourself in need of a second implementation it is pretty easy anyway to separate the interface from the implementation.

Reply:
And the difference between `search()` and e.g. `requestCalendarPermission()` is that `search` creates a `Job` that might capture that context and this way the reference may outlive the activity, while `requestCalendarPermission` doesn't?

I've now changed it to a new module in `core`, thanks for the thorough explanation. Still, wouldn't it make sense for `SearchVM` to implement some disposable interface that then cancels the `Job`? (Java doesn't know `IDisposable` like C#, but kotlin isn't java, amirite?)

> interface UserLocationProvider and a separate implementation class UserLocationProviderImpl: UserLocationProvider

I've stuck with a class with an internal constructor now.

> joke's on me who has written like two unit tests in total for this repo. I will add them eventually, I promise!

Testing in production also works I suppose 🤠

Reply:
> And the difference between search() and e.g. requestCalendarPermission() is that search creates a Job that might capture that context and this way the reference may outlive the activity, while requestCalendarPermission doesn't?

The reason `requestCalendarPermission` has a context parameter is that you need an activity context to launch permission requests. I would prefer it if I could call PermissionsManager.requestPermission without a context parameter, but that doesn't work in Android.

`search` on the other hand isn't tied to any activity and we can easily use the application context every time we need one. And since Koin knows that application context, it can inject it and we don't need to worry about it.

But yes, `search` potentially outliving the activity is definitely another reason why we should avoid it here.